AVM2 interpreter

kmeisthax opened the pull request:
This implements an AVM2 interpreter. AVM2 is the virtual machine that ActionScript 3.0 programs compile into; it's completely separate from AVM1 and almost became part of the web platform itself in the form of ECMAScript 4 (no published standard exists) and Adobe Tamarin (FOSS but MPL trilicensed instead of Apache).

Notable features include packages, namespaces, access control, E4X, type assertions, a very explicit scope stack, no primitive values (if the one ES4 overview PDF I could find is accurate), and more. Coming from modern JavaScript, or even AS2; this feels like a completely different programming language and it took a while to understand very basic things about the VM. (This is probably why non-Flash web developers outright rejected ES4.) Nevertheless, it's part of Flash, so we need to implement it as it gates a whole bunch of interesting content.

# In this PR

 * [x] `Avm2` interpreter state
 * [x] Values, objects, activations, functions, namespaces, properties, and slots
   * [x] Interpreter uses class and trait definitions separate from the ones in `swf`
   * [x] Interfaces, including their effect on `istype` instructions
 * [x] Trait instantiation
   * [x] Re-do trait instantiation because traits exist on instance objects directly
   * [x] Make sure traits actually don't get instantiated on prototypes
 * [x] Control flow instructions
   * [x] `jump`
   * [x] `iftrue`
   * [x] `iffalse`
   * [x] `ifstricteq`
   * [x] `ifstrictne`
   * Other control flow instructions are future work
 * [x] Object iteration
 * [x] `Object`, `Function` builtins
 * [x] `Class` class
 * ~~AVM2 `StageObject`~~ - see future work
 * ~~`SymbolClass` tag~~ - see future work
 * ~~`FileAttributes` tag & VM isolation~~ - see future work
 * [x] Tests
  * We can't implement `Function.prototype.apply` without working `Array`s, which is outside the scope of this PR
  * We can't test closures because the test I wrote for them uses `coerce`.

# PR Roadmap
The following are features out of scope of this PR. Due to the fact that this PR's scope has more or less been completed, I'm working on them in separate branches. If they're relatively "complete" before this PR is merged in, they may be merged in.

 * Primitive type coercion, abstract comparisons, and abstract equality -  `avm2-primitives` on my user's repo.
   * Numerical calculations - some implementations exist on the above but will be given their own grandchild PR
   * String manipulation - not yet started
   * Primitive boxing - not yet started
 * Native builtins - `avm2-nativeclasses` on my user's repo. Covers defining ES4 classes in Rust.
   * AVM2 `StageObject` - `TObject` implementation that holds stage objects, plus native implementations of the builtin display object classes.
     * `SymbolClass` tag - required for executing frame scripts
     * `FileAttributes` tag & VM isolation - Already implemented in #661, but for the purpose of warning users about our lack of AVM2 support. We will need to change this to configuring what VM a particular movie's display objects belong to.

kmeisthax commented:
I am releasing the draft lock on this PR and requesting review and merge. Hopefully it passes CircleCI.

I'm deliberately walling off a whole bunch of future AVM2 features for future PRs, as mentioned in the PR description. This PR already has 143 commits and makes `object_prototyping`, a PR I had to cut up into four painful-to-rebase pieces, look *minimal*. I've specifically selected future work that isn't likely to have interdependencies between future PRs, and thus can be developed separately on top of the work I've done bootstrapping a basic AVM2 interpreter.

There are approximately 31 tests, a significant number of which challenged my notions about how this interpreter should work. I think I at least have the object model locked, though I wasn't able to write a test that uses slots. In fact, `clippy` seems to think `Slot::Occupied` is unreachable, even though there's a function constructing it. I disabled that warning for now.

Justin-CB commented:
Is there any ETA on this pull request?

jfmherokiller commented:
> Is there any ETA on this pull request?

Out of my own curositiy i tried to merge this into master (current as of today) and it seems this branch is way behind. Just based off a cursory observation (im no rust master) we might need to create an avm super type that encases both avm1 and avm2 and cast based upon the information extracted from the swf.

kmeisthax commented:
@jfmherokiller There's no need to do that, I just need to resolve the merge conflicts. AVM1 and AVM2 are supposed to be entirely separate; the conflicts are in other parts of the player where I added AVM2 stuff to.

kmeisthax commented:
I should also point out that this does not contain the capability to play AS3 *games*. It can't even add numbers yet - it's purely the AVM2 interpreter loop, object model, and just enough builtins to get tests to run.

jfmherokiller commented:
the question then i have is how will the player distinguish between an avm1 and avm2 swf?

kmeisthax commented:
As you can tell by that massive list of new commits I have rebased the AVM2 PR and resolved the merge conflicts.

Given @Dinnerbone's recent work with init actions I'm wondering if ABC files should also execute immediately rather than resolving on the action queue.

jfmherokiller commented:
I will say I did succed in adding a kind of standin for FileAttributes (I just started using rust today so code will be very basic) in movie_clip.rs.
[mydiff.txt](https://github.com/ruffle-rs/ruffle/files/4507396/mydiff.txt)

kmeisthax commented:
So, that's great, but kinda out of scope for this PR. I specifically excluded anything to do with either `FileAttributes`, or tying display objects into the AVM2 interpreter, because this PR is already fairly sprawling and I want @Herschel's eyes on it and this merged before I start doing other related work.

Whenever this actually gets merged, and we start work on an "AVM2 DisplayObjects" PR, then we'd want to start parsing `FileAttributes` so we'd know how to instantiate a given movie. The relevant flags would both have to set a bit on the containing `SwfMovie` (so that contained movieclips can't mix VMs) and instantiate the entire movie hierarchy as either AVM1 or AVM2. Furthermore, I'm a little wary of parsing `FileAttributes` in preload because there are script actions and instantiations that can happen *before* a `FileAttributes` tag has been encountered which *do* need to know what VM is allowed to run for a given movie. Ideally, we'd scan the SWF's root movie clip *before* `preload` (or even before making the initial `MovieClip`) so we could get at those attributes and use that to instantiate the correct VM's `Object`.

kmeisthax commented:
Turns out this thing had been sitting unmergeable for a good week or two 'cause I didn't notice the fact that beta/nightly clippy tests are now being run.

Also, given that #675 landed I'm now considering adjusting the AVM2 `Value` and `Object` to match...

kmeisthax commented:
I've implemented a handful of refactors @Dinnerbone did on the AVM1 side to the AVM2 PR, and I've also changed the way class loading works to make native classes far easier to implement. Aside from that, I don't foresee much else making it to this PR, unless I discover something foundationally wrong with how AVM2 is currently implemented, or another refactor hits AVM1 that I feel would be valuable in AVM2.

AS3 has a significantly larger API surface than AS2 did, and was specifically designed to support more complex programs. Currently, our AS2 support is getting pretty good, and the average untested game at least has some functionality under Ruffle. It will likely take the same amount of time to get AS3 movies to the same level as AS2 currently is, with the additional caveat that there will be far less partial progress as we get there.

Since this PR is already extremely large, I'm currently working on two child branches of this PR:

- `avm2-primitives`: ECMA-262 compliant coercions, conversions, and abstract comparison operators. Will enable AVM operations relating to numerical calculations.
- `avm2-nativeclasses`: Adding `Class` entries for existing defined builtins. Will enable defining built-in AS3 classes in Rust, and eventually, classes on DisplayObjects (which will, presumably, allow some games to work).

fp07 commented:
This looks great, what is the expected ETA so we can test this on nightly versions? Hopefully there is an good AS3 support before the end of this year.

Dinnerbone commented:
In case you missed it, it doesn't compile atm!

Dinnerbone commented on core/src/avm2/names.rs:
```diff
@@ -0,0 +1,309 @@
+#[derive(Clone, Debug, Collect)]
+#[collect(no_drop)]
+pub struct Multiname {
+    ns: Vec<Namespace>,
+    name: Option<String>,
```
Is it possible for these to be Cows? We seem to construct them frequently based on existing data

kmeisthax replied:
Namespaces are now `AvmString`s, so I believe there is no further need for deduplication as a namespace is just a string reference and an enum tag now.

Dinnerbone commented on core/src/avm2/names.rs:
```diff
@@ -0,0 +1,309 @@
+#[derive(Clone, Collect, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[collect(no_drop)]
+pub enum Namespace {
+    Namespace(String),
+    Package(String),
+    PackageInternal(String),
+    Protected(String),
+    Explicit(String),
+    StaticProtected(String),
+    Private(String),
```
Same here, is it possible for these to be Cows? We seem to construct them frequently based on existing data

kmeisthax replied:
Unfortunately, no. For various reasons, `Value` needs to be able to hold a `Namespace`, and if I stick a `Cow` in here then the lifetime I use has to leak out into everything that works with namespaces. I tried to work around this with `Namespace<'static>`, which I could fit into a `Value<'gc>`, but then implicit lifetimes broke in anything that needs to hold references to a `Value`, which turns out to be a lot of code (it's part of the definition of native methods/functions). This also broke in the same way with `Namespace<'gc>`, a lifetime bound which makes absolutely no sense.

Since there isn't a way to extinguish the borrowed half of the `Cow` when I specifically need to structurally require an owned value, I tried moving this up one notch by implementing a separate borrowed namespace type called `NS`. The idea was that you could hold a `Cow<'a, NS<'a>>`, and then have it return a `Namespace`. This also did not work, due to a chain of conflicting requirements:

 * In order to stick a type in a `Cow`, it needs to implement `ToOwned`.
 * In order to implement `ToOwned`, your type cannot be `Clone` (otherwise you get an auto implementation), which I was able to work around
 * In order to implement `ToOwned`, the type you do return must implement `Borrow`.
 * In order to implement `Borrow`, your owned type must return a reference to the borrowed type... or, `&'a NS<'a>`, which makes no sense for an enum being constructed on-stack.
 * Ergo, we can't make a pair of types that `Cow` can work with.

If you have any other suggestions about how to be able to pass borrowed-string namespaces, please let me know.

kmeisthax replied:
I straight up replaced all the strings in AVM2 with `AvmString`, which should provide the same string deduplication opportunities as `Cow<'static, str>` does. I'm going to mark that as a resolution to this concern.

Dinnerbone commented on core/src/avm2/value.rs:
```diff
@@ -0,0 +1,287 @@
+        if let Value::Object(object) = self {
+            Ok(*object)
+        } else {
+            Err(format!("Expected Object, found {:?}", self).into())
```
It seems to be only `null` and `undefined` should error out here, the rest should box. I guess that's coming in a later PR?

kmeisthax replied:
That will be in the `avm2-primitives` branch, a child branch of this one. That currently has most of the coercions and abstract equality, but I still need to implement abstract comparison and value boxing.

Dinnerbone commented on core/src/avm2/value.rs:
```diff
@@ -0,0 +1,287 @@
+    }
+
+    /// Coerce a value into a string.
+    pub fn coerce_string(self) -> String {
```
Can this be a Cow?

Dinnerbone commented on core/src/avm2/value.rs:
```diff
@@ -0,0 +1,287 @@
+
+/// Retrieve a string from an ABC constant pool, yielding `""` if the string
+/// is the zero string.
+pub fn abc_string(file: &AbcFile, index: Index<String>) -> Result<String, Error> {
```
Can this be `&str`?

Dinnerbone commented on core/src/avm2/value.rs:
```diff
@@ -0,0 +1,287 @@
+/// should cause the runtime to halt. `None` indicates that the zero string is
+/// in use, which callers are free to interpret as necessary (although this
+/// usually means "any name").
+pub fn abc_string_option(file: &AbcFile, index: Index<String>) -> Result<Option<String>, Error> {
```
Can this be `Result<Option<&str>, Error>`?

Dinnerbone commented on core/src/avm2/activation.rs:
```diff
@@ -0,0 +1,1627 @@
+        &self,
+        method: Gc<'gc, BytecodeMethod<'gc>>,
+        index: Index<String>,
+    ) -> Result<String, Error> {
```
Can this be `Result<&str, Error>`?

kmeisthax replied:
There isn't a suitable lifetime for the returned string; the AbcFile that the string comes from is owned by the `Gc<'gc, BytecodeMethod<'gc>>` and we can only get temporary borrows for it.

Herschel replied:
I have not looked over everything thoroughly, but you can use `std::cell::Ref::map`, e.g. starting in `TranslationUnit`:
```
    pub fn abc_ref(&self) -> std::cell::Ref<AbcFile> {
        std::cell::Ref::map(self.0.read(), |tx| &*tx.abc.0)
    }
```
and then similarly in `BytecodeMethod`:
```
    pub fn abc_ref(&self) -> std::cell::Ref<AbcFile> {
        self.txunit.abc_ref()
    }
```
and so forth, `Activation::pool_string` takes `&BytecodeMethod` instead of a `Gc`. Not sure if this is a helpful design, but at least its possible in theory.

kmeisthax replied:
That actually worked! I had to do some value calisthenics to get around the fact that there is no `Ref::map_result`, though. This code looks super verbose now:

```
let mut maybe_error = None;
let maybe_ref = Ref::map(method.abc_ref(), |abc| {
    match value::abc_string(abc, index) {
        Ok(s) => s,
        Err(e) => {
            maybe_error = Some(e);
            ""
        }
    }
});

if let Some(err) = maybe_error {
    Err(err)
} else {
    Ok(maybe_ref)
}
```

kmeisthax replied:
Thanks to Dinnerbone's `AvmString` work this approach is actually obsolete anyway, so I'm going to mark this as resolved.

Dinnerbone commented on core/src/avm2/activation.rs:
```diff
@@ -0,0 +1,1627 @@
+        register_name: Index<String>,
+        register: u8,
+    ) -> Result<FrameControl<'gc>, Error> {
+        if is_local_register {
+            let register_name = self.pool_string(method, register_name)?;
+            let value = self.local_register(register as u32)?;
+
+            avm_debug!("Debug: {} = {:?}", register_name, value);
+        } else {
+            avm_debug!("Unknown debugging mode!");
+        }
```
We can wrap this whole thing behind `avm_debug` feature flag, and save the lookups.

Dinnerbone commented on core/src/avm2/activation.rs:
```diff
@@ -0,0 +1,1627 @@
+        method: Gc<'gc, BytecodeMethod<'gc>>,
+        file_name: Index<String>,
+    ) -> Result<FrameControl<'gc>, Error> {
+        let file_name = self.pool_string(method, file_name)?;
```
Same here, no need to waste cycles.

Herschel commented on core/src/avm2/script.rs:
```diff
@@ -0,0 +1,286 @@
+        mc: MutationContext<'gc, '_>,
+    ) -> Result<Method<'gc>, Error> {
+        let write = self.0.write(mc);
+        if let Some(method) = write.methods.get(&method_index) {
+            return Ok(method.clone());
+        }
+
+        drop(write);
+
+        let method: Result<Gc<'gc, BytecodeMethod<'gc>>, Error> =
+            BytecodeMethod::from_method_index(self, Index::new(method_index), mc)
+                .ok_or_else(|| "Method index does not exist".into());
+        let method: Method<'gc> = method?.into();
+
+        self.0
+            .write(mc)
+            .methods
+            .insert(method_index, method.clone());
```
Feels like we should be able to use the entry API with `or_insert_with`, although I guess the issue is that `from_method_index` does some `txunit.read()`.

kmeisthax replied:
The timing of when `GcCell` locks are held and released is important here (which is why I used the "wrapped `GcCell` pattern" here). Due to the way traits work, all four of these methods can recurse into one another on the same stack at any time, so I'm being *highly* conservative with how long locks are actually being held.

`BytecodeMethod` *never* needs a write lock on the translation unit, and it only ever reads it to get an ABC. In fact, it reads it three times too many because this used to be `abc_ref` (which caused other double-borrow problems). I'm going to change this particular constructor into taking both a `TranslationUnit` and an `AbcFile`, so it doesn't need to hold either read or write locks and we can use entries.

kmeisthax replied:
Turns out that won't work, but not for the reason you'd think: `or_insert_with` is infallible. We need to generate a verification error if the method index doesn't exist, which we can't return inside of the `or_insert_with` callable. Ergo, I'll be leaving this as-is.

(This seems to be a problem with a lot of Rust methods that take closures...)

Herschel commented on core/src/avm2/script.rs:
```diff
@@ -0,0 +1,286 @@
+    ) -> Result<AvmString<'gc>, Error> {
+        Ok(self
+            .pool_string_option(string_index, mc)?
+            .unwrap_or_else(|| "".into()))
```
unwrap_or_default

kmeisthax replied:
Huh. That seems to be fine now. I seem to recall `unwrap_or_default` tripping Clippy lints at some point...

kmeisthax commented on core/src/avm2/names.rs:
```diff
@@ -0,0 +1,335 @@
+        multiname_index: Index<AbcMultiname>,
+        mc: MutationContext<'gc, '_>,
+    ) -> Result<Self, Error> {
+        let actual_index = multiname_index.0 as usize - 1;
```
This method is only called to resolve names during trait instantiation, so this should generate a runtime error rather than an underflow panic. I will change it as such.

kmeisthax commented on core/src/avm2/method.rs:
```diff
@@ -0,0 +1,179 @@
+    fn trace(&self, cc: CollectionContext) {
+        match self {
+            Method::Native(_nf) => {}
+            Method::Entry(a2me) => a2me.trace(cc),
```
"AVM2 Method Entry" though I'll just change this to `entry` to avoid confusion.

kmeisthax commented on core/src/avm2/script_object.rs:
```diff
@@ -0,0 +1,907 @@
+    /// The class that this script object represents.
+    class: ScriptObjectClass<'gc>,
+
+    /// Enumeratable property names.
+    enumerants: Vec<QName<'gc>>,
```
So, confession time: I've tried to read avmplus source as little as possible due to it being a different license from our own (we're Apache, they're MPL). I've been going entirely off of the AVM2 overview and AS3 language reference documents. When I read the spec, I figured there would be only a few ways to actually implement it:

- Add a new parameter space for enumeration indices and their resulting names
- Hand hash indices directly to the program (not possible until this is stabilized in Rust: https://github.com/rust-lang/rust/issues/56167)
- Do some special-cased nonsense that disregards the index values entirely and breaks if a program tries to probe index space.

I'm both not surprised and yet still disappointed that Adobe chose the latter in avmplus :/

kmeisthax commented on core/src/avm2/script.rs:
```diff
@@ -0,0 +1,286 @@
+
+#[derive(Copy, Clone, Debug, Collect)]
+#[collect(no_drop)]
+pub struct TranslationUnit<'gc>(GcCell<'gc, TranslationUnitData<'gc>>);
```
So, the name came about while I was switching the AVM2 interpreter away from pulling ABC traits and data directly to storing them in other structures that could be dynamically created (which we need for `avm2-nativeclasses`.) I can't call it `AbcFile` because that's already in use by `swf`, so `TranslationUnit` (a term used by C/C++ compiler specifications to refer to a single source file) seemed like a good option, since, as far as I could tell I was getting one `DoABC` per source file in my tests...

If you have a better option I'll definitely change the name.

kmeisthax commented on core/src/avm2/value.rs:
```diff
@@ -0,0 +1,262 @@
+/// TODO: AVM2 also needs Scope, Namespace, and XML values.
+#[derive(Clone, Collect, Debug)]
+#[collect(no_drop)]
+pub enum Value<'gc> {
+    Undefined,
+    Null,
+    Bool(bool),
+    Number(f64),
+    String(AvmString<'gc>),
+    Namespace(Namespace<'gc>),
+    Object(Object<'gc>),
+}
```
I have no strong opinion either way.

I did use `Value` specifically because it matches the AVM1 `Value`. I doubt we're going to have people used to hacking on avmplus wondering where `Atom` is here, as much as we're going to have people used to hacking on Ruffle's AVM1 interpreter wondering where `Value` is.

kmeisthax commented on core/src/avm2/property.rs:
```diff
@@ -0,0 +1,260 @@
+/// TODO: Replace with AVM2 properties for traits
+#[derive(EnumSetType, Debug)]
+pub enum Attribute {
+    DontDelete,
+    ReadOnly,
```
Should I restore the flag now or when we actually encounter a `DontEnum` property in AVM2 code?

Herschel commented:
Ok, let's get this in! Thanks for the patience while this PR sat around for a long time. I'm excited to make progress here.

(BTW, I meant to say that I *don't* feel strongly about the name stuff in my earlier post, sorry if that came off the wrong way).
